### PR TITLE
[TASK] Refactor widget architecture

### PR DESCRIPTION
**Pull request by NeoBlack**
- Introduce multiple interfaces for different widget scopes
- cleanup the abstract classes
- add a short introduction to interfaces and abstract classes

### CODE REVIEW

**Comment by haassie** [comment]
Can you please pull master in this branch and resolve the merge conflict?

**Comment by NeoBlack** [comment]
rebased and pushed again

File: Classes/Widgets/AbstractChartWidget.php [hunk 1]
```diff
@@ -3,64 +3,66 @@
...
-    protected function prepareChartData(): void
+    public function getEventData(): array
     {
-        // This method need to be implemented by the final class
+        $this->eventData['graphConfig'] = [
+            'type' => $this->chartType,
+            'data' => $this->chartData,
+            'options' => $this->chartOptions
+        ];
+        $this->prepareChartData();
```
Comment (haassie) [inline Classes/Widgets/AbstractChartWidget.php]: I guess this method should be called before the `graphConfig` is set. Now the `data` element is empty.

File: Classes/Widgets/SysLogErrorsWidget.php [hunk 2]
```diff
@@ -101,4 +65,17 @@ protected function getNumberOfErrorsInPeriod(int $start, int $end): int
             ->execute()
             ->rowCount();
     }
+
+    protected function calculateDataForLastDays(int $days): array
+    {
+        $data = [];
+        for ($daysBefore=$days; $daysBefore--; $daysBefore>0) {
+            $labels[] = date('d-m-Y', strtotime('-' . $daysBefore . ' day'));
```
Comment (haassie) [inline Classes/Widgets/SysLogErrorsWidget.php]: `$labels` isn't returned so this results in empty labels now.

File: Classes/Widgets/AbstractChartWidget.php (hunk 1 above)
Comment (NeoBlack) [reply Classes/Widgets/AbstractChartWidget.php]: correct, will push an update

**Comment by NeoBlack** [comment]
pushed the update for the open issues

**Pull request merged**